### PR TITLE
feat: pass secrets to Container.Build and Directory.Build

### PR DESCRIPTION
**Pull request by dolanor**
Now you can pass secrets to a Dockerfile build like in docker with

```Dockerfile
RUN --mount=type=secret,id=xxx,dst=/run/secrets/xxx
```

```
docker build --secret id=xxx,src=/my/file .
```

You can use

```graphql
type Container {
    build(dockerfile: [String], platform: [Platform], buildArgs: [BuildArg!], target: String, secrets: [SecretID!]): Container!
}

type Directory {
    dockerBuild(dockerfile: [String], platform: [Platform], buildArgs: [BuildArg!], target: String, secrets: [SecretID!]): Container!
}
```

## Limitations

It will mount in `/run/secrets/<secret-id>`, you cannot decide the target mount point with this current API.
It could be improved later on and discussed on how the API should look like.

### CODE REVIEW

File: core/secret.go [hunk 1]
```diff
@@ -40,9 +39,7 @@ func NewSecretFromHostEnv(name string) (*Secret, error) {
 type SecretID string
 
 func NewSecretID(name, plaintext string) (SecretID, error) {
-	digestBytes := sha256.Sum256([]byte(plaintext))
```
Comment (marcosnils) [inline core/secret.go:43]: Is this change intentional? we're not using `plaintext` anymore in this function?

File: core/integration/container_test.go [hunk 2]
```diff
@@ -261,6 +261,24 @@ CMD echo "stage2"
...
+		sec := c.SetSecret("my-secret", "barbar")
+		require.NoError(t, err)
+
+		src := contextDir.
+			WithNewFile("Dockerfile",
+				`FROM golang:1.18.2-alpine
+WORKDIR /src
+RUN --mount=type=secret,id=my-secret test "$(cat /run/secrets/my-secret)" = "barbar"
+RUN --mount=type=secret,id=my-secret cp /run/secrets/my-secret  /secret
+CMD cat /secret
```
Comment (marcosnils) [inline core/integration/container_test.go:275]: nit: improve this assertion to `CMD test $(cat /secret) == "barbar"`. 
We can then assert the go test with err == nil instead of "***"

File: core/secret.go (hunk 1 above)
Comment (dolanor) [reply core/secret.go:43]: It was a way to get the cache busted when the plaintext of a secret gets changed, but it leaked some info (the secret has changed), and it made the support in Dockerfile painful to users (they had to use some templated Dockerfile, with the secret ID already encoded with the digest of the plaintext).
Here it's more compatible with simple secret ID for buildkit. The secret name becomes the secret ID. `setSecret` will just register the secret name/ID to its plaintext in the `secret.Store` buildkit session attachable.

File: core/integration/container_test.go (hunk 2 above)
Comment (dolanor) [reply core/integration/container_test.go:275]: Since I already do this on line 273, I feel it's a repeat that would just confirm that the `cp` from the secret to another file works instead of also checking the scrubbing.

Comment (marcosnils) [reply core/integration/container_test.go:275]: > Since I already do this on line 273, I feel it's a repeat that would just confirm that the `cp` from the secret to another file works instead of also checking the scrubbing.

you're correct. Totally missed this.

File: core/secret.go (hunk 1 above)
Comment (marcosnils) [reply core/secret.go:43]: got it, thx! since we're not using `plaintext` anymore in that function, could we remove it? or that breaks some user facing API's?

File: core/schema/container.graphqls [hunk 3]
```diff
@@ -63,6 +63,9 @@ type Container {
 
     "Target build stage to build."
     target: String
+
+    "Secrets to pass to the build."
```
Comment (TomChv) [inline core/schema/container.graphqls:67]: Maybe add a small sentence about the `/run/secrets` so the user knows where the secret will be from the doc.

File: core/schema/directory.graphqls [hunk 4]
```diff
@@ -190,6 +190,9 @@ type Directory {
 
     "Target build stage to build."
     target: String
+
+    "Secrets to pass to the build."
+    secrets: [SecretID!]
```
Comment (TomChv) [inline core/schema/directory.graphqls:195]: Maybe add a small sentence about the `/run/secrets` so the user knows where the secret will be from the doc.

**Review by TomChv — APPROVED** [review]
LGTM, left comments

File: core/schema/directory.graphqls (hunk 4 above)
Comment (dolanor) [reply core/schema/directory.graphqls:195]: Very good point, thanks.

File: core/schema/container.graphqls (hunk 3 above)
Comment (dolanor) [reply core/schema/container.graphqls:67]: Very good point, thanks.
